### PR TITLE
Improve error message

### PR DESCRIPTION
**Pull request by madoleary**
## Self proof reading checklist

- [ ] Did I run Pronto and fixed all warnings
- [x ] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [x ] Did I test the right thing?
- [x ] Did I test corner cases, non happy path (defensive testing)?
- [ ] Check that I used the ad-hoc patterns and created files accordingly
- [ ] Check code efficiency with record intensive project
- [ ] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

File: lib/orbf/rules_engine/data/contract.rb [hunk 1]
```diff
@@ -1,17 +1,46 @@
...
+        keys_to_validate.each do |key|
+          if values[key] && (values[key].nil? || values[key].blank?)
+            missing_attrs << key
+          else 
+            values.fetch(key) { missing_attrs << key } 
+          end 
+        end
+    
+        if missing_attrs.any?
+          raise ContractValidationException.new(missing_attrs, values.fetch("org_unit") { nil })
```
Comment (pjaspers) [inline lib/orbf/rules_engine/data/contract.rb:42]: We don't really care about the `values.fetch("org_unit")` here, so we could do `values["org_unit"]` which would do the same as the fetch with a default `nil`

File: lib/orbf/rules_engine/data/contract.rb [hunk 2]
```diff
@@ -1,17 +1,46 @@
 module Orbf
   module RulesEngine
+    class ContractValidationException < StandardError
+      def initialize(missing_attrs, org_unit)
+        org_unit_missing = missing_attrs.include?("org_unit")
```
Comment (pjaspers) [inline lib/orbf/rules_engine/data/contract.rb:5]: If we make this `org_unit ||= {}` we don't have to do the `!org_unit_missing` check, since the `org_unit.fetch("name") {"missing_org_unit_name"}` will return `missing_org_unit_name`.

And then we can make the message always be `"The following attributes are missing for this contract: #{missing_attrs}" (orgunit_name - #{org_unit_name || 'missing'},orgunit_ id - #{org_unit_id || 'missing')

So we lose the ternary (mostly because I don't like ternaries, but also because it makes the code a little bit easier to read?)

File: lib/orbf/rules_engine/data/contract.rb [hunk 3]
```diff
@@ -1,17 +1,46 @@
...
         @org_unit = field_values.fetch("org_unit")
         @start_period = field_values.fetch("contract_start_date").gsub("-", "").slice(0, 6)
         @end_period = field_values.fetch("contract_end_date").gsub("-", "").slice(0, 6)
         @calendar = calendar
+      end 
+
+      def validate!(values)
+        missing_attrs = []
+        keys_to_validate = ["id", "contract_start_date", "contract_end_date", "org_unit"]
+        keys_to_validate.each do |key|
```
Comment (pjaspers) [inline lib/orbf/rules_engine/data/contract.rb:33]: I'd go for an `inject` here, but that's mostly just a style thing, I know Stephan is also more a fan of a simple each and filling a variable :)

File: spec/lib/orbf/rules_engine/data/contract_spec.rb [hunk 4]
```diff
@@ -133,6 +133,117 @@ def test_overlaps(other_start, other_end, expected)
...
+          field_values = {
+            "id"                  => "GPAGzdsLDTP",
+            "date"                => "2018-06-01T00:00:00.000+0000",
+            "contract_end_date"   => "2021-12-31",
+            "contract_type"       => "PMA"
+          }
+      
+          missing_attrs = "contract_start_date, org_unit"
+          
+          expect { Orbf::RulesEngine::Contract.new(field_values, calendar) }.to raise_error(Orbf::RulesEngine::ContractValidationException)
```
Comment (pjaspers) [inline spec/lib/orbf/rules_engine/data/contract_spec.rb:149]: Again, personal thing here, but I'll explain my reasoning!

I don't like multiple `expect` in a single test because they tend to be more brittle. If we change the Class of the error we now have to update it in multiple tests. So I would write one test to check that it raises the Error we expect, and then in the other tests it would rely on raise_error.with_message (since I don't care about the error class in those tests, just that it matches the message)

Since we've moved some logic into the error class (always need to be careful when doing that, because if we make an error inside the Error, we lose our nice error), I would test the Error instead of the `Contract` class.

Like

``` ruby
describe ContractValidationException do
   it "has custom messages on missing org unit attributes" do
      exception = ContractValidationException .new(["org_unit"], {"name" => "", "id" => ""})
      
      except(exception.message).to match(/#{relevant_org_unit_id}/)
  end
end
```

And the in the contract spec, I'd just check that it raises the error like we expect. That way if we modify the behavior of the Error we only need to update the specs for that error.

Again, all this is more a personal style thing (but imho makes for less brittle tests), so I'm good with the code as is, I think we're doing exactly what we wanted it to do.

**Review by pjaspers — APPROVED** [review]

File: spec/lib/orbf/rules_engine/data/contract_validation_exception_spec.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    missing_attrs = ["contract_start_date"]
+    org_unit = { "id" => "cLc2uthCRfm", "name" => nil, "path" => "/pL5A7C1at1M/BmKjwqc6BEw/DkK8tVZ6xfJ/z6sJvc6NR59/cLc2uthCRfm" }
+    exception = Orbf::RulesEngine::ContractValidationException.new(missing_attrs, org_unit)
+    
+    expect(exception.message).to match(/#{"name - missing"}/)
+  end
+
+  it "has custom messages on missing org unit" do
+    missing_attrs = ["contract_start_date"]
+    exception = Orbf::RulesEngine::ContractValidationException.new(missing_attrs, org_unit=nil)
```
Comment (pjaspers) [inline spec/lib/orbf/rules_engine/data/contract_validation_exception_spec.rb:12]: I have never thought of doing a `org_unit=nil` as a paremeter to be passed, makes sense that it works :)

**Review by pjaspers — APPROVED** [review]
Looks good to me!
